### DISCUSSION TITLE
关于SDK的一些问题(Retry/Provider/Client)

### DISCUSSION THREAD

**Discussion by northseadl** [ArtisanCloud/PowerWeChat#214] [General]
## 重试机制
目前的Token失效检查检查机制依赖于Middleware, Middleware的接口机制导致重试没有办法实现的完美, 要实现Token刷新重试, 需要在一个作用域里能访问Request和Response, 有以下三个想法:
1. 实现洋葱模型中间件
  ```
    Middleware(request, response) {
      // pre request
     
      handle(request, response)
      
      // after request
    }
  ```
2. 修改目前的中间件接口参数, 传递 context, 将 Request 附加到 context 的 Value 中
3. 实现AuthClient, 实现 Auth Roundtripper, 在 Roundtripper 中 附加 Token 并实现 刷新Token&重试 机制, 该实现方式 gout 层是无感的

## Client 的内嵌方式
当前的 Provider 中的 Client 为内嵌方式, 这导致在 Api 层能够访问到 Client 方法, 直观体验是 Ide 的提示里很乱

## 企业微信的 Oauth Provider
这个 Provider 的 BaseClient 调用的是依赖包里的 Base, 几乎没有实现 Token 刷新机制,  WeComApp.OAuth.Provider 中的方法几乎都受到此影响

**Comment by Matrix-X** [comment]
谢谢你的讨论，你的想法很好，其实我们也有预留在middleware的retry机制，但是后来发现，我们使用gout的中间键机制里，只有在非200的情况下才能捕获。。。

所以暂时我们先在每次判断业务请求后，判断是否会是4001等，从而决定是否要refresh token。

正如你所设想，这个middleware的确是我想要的方案，请加我的企业微信，我们可以讨论一下，怎么优化这个机制。

**Reply by northseadl** [reply]
已发送添加申请
